### ISSUE TITLE
Using Concatenate in return type causes wrong parameter inference.

### ISSUE THREAD

**Issue by MHDante** [microsoft/pyright#1455]
**Describe the bug**
Usung `Concatenate` in a return value type annotation yields the wrong type. I think it's not doing concatenation on the return paramspec.

**To Reproduce**
Create a function that takes a Callable that decomposes the type using Concatenate and then recomposes it in the return type hint.

**Expected Behaviour**
The below error messages should not be reported.

**Screenshots or Code**

``` Python
from typing import Callable, Protocol, TypeVar
from typing_extensions import Concatenate, ParamSpec
from threading import RLock
import functools

class HasLock(Protocol):
    _lock : RLock

S = TypeVar("S", bound=HasLock)
P = ParamSpec("P")
R = TypeVar("R")

def with_lock(func: Callable[Concatenate[S, P], R]) -> Callable[Concatenate[S, P], R]:

    @functools.wraps(func)
    def wrapper(self: S, *args : P.args, **kwargs : P.kwargs)->R:
        with self._lock:
            return func(self, *args, **kwargs)

    return wrapper


class MyClass:
    def __init__(self):
        self._lock = RLock()

    @with_lock
    def test_1(self, cheem:int)-> str:
        return f"cheem: {cheem}"

    @with_lock
    def test_2(self)-> str:
        return f"no cheems"

@with_lock
def test_3(cls : MyClass, cheemz: int) -> str:
    return f"cheemz: {cheemz}"


testClass = MyClass()

res1 : str = testClass.test_1(42)

res2 : str = testClass.test_2()

res3 : str = test_3(testClass, 42)

res4 : Callable[[MyClass, int], str] =  with_lock(test_4)

```

![image](https://user-images.githubusercontent.com/4652122/107130173-f46c7b80-6898-11eb-8c84-b0168f0f81db.png)

![image](https://user-images.githubusercontent.com/4652122/107130446-8d03fb00-689b-11eb-86b7-3fa64ba98023.png)

![image](https://user-images.githubusercontent.com/4652122/107130452-9725f980-689b-11eb-923d-b649967afd96.png)

![image](https://user-images.githubusercontent.com/4652122/107130609-3a2b4300-689d-11eb-83d1-2eb14103bda8.png)

**VS Code extension or command-line**
ms-python.vscode-pylance v2021.2.1-pre.1

**Additional context**
You guys are super cool.

**Comment by erictraut** [comment]
Thanks for the detailed bug report. This will be fixed in the next release.

If you're curious, here's the PR for the fix: https://github.com/microsoft/pyright/commit/8d8c63065a9c28fd3c1d4dc7c2d7a49972976755

I based the unit test on your sample code. :)

**Comment by erictraut** [comment]
This is now fixed in Pyright 1.1.109, which I just published. It will also be in the next release of Pylance.
